api/ipfsproxy/headers.go

```diff
@@ -0,0 +1,157 @@
+var extractHeadersDefault = []string{
+	"Access-Control-Expose-Headers",
+}
+
+// ipfsHeaders returns all the headers we want to extract-once from IPFS: a
+// concatenation of extractHeadersDefault and config.ExtractHeadersExtra.
+func (proxy *Server) ipfsHeaders() []string {
+	return append(extractHeadersDefault, proxy.config.ExtractHeadersExtra...)
+}
+
+// rememberIPFSHeaders extracts headers and stores them for re-use with
+// setIPFSHeaders.
+func (proxy *Server) rememberIPFSHeaders(hdrs http.Header) {
+	for _, h := range proxy.ipfsHeaders() {
+		proxy.ipfsHeadersStore.Store(h, hdrs[h])
+	}
+}
+
+// rememberIPFSHeaders adds the known IPFS Headers to the destination
+// and returns true if we could set all the headers in the list.
+// False is used to determine if we need to make a request to try
+// to extract these headers.
+func (proxy *Server) setIPFSHeaders(dest http.Header) bool {
+	r := true
+	for _, h := range proxy.ipfsHeaders() {
+		v, ok := proxy.ipfsHeadersStore.Load(h)
+		if !ok {
+			r = false
+			continue
+		}
+		dest[h] = v.([]string)
+	}
+	return r
+}
+
+// copyHeadersFromIPFSWithRequest makes a request to IPFS as used by the proxy
+// and copies the given list of hdrs from the response to the dest http.Header
+// object.
+func (proxy *Server) copyHeadersFromIPFSWithRequest(
+	hdrs []string,
+	dest http.Header, req *http.Request,
+) error {
+	res, err := proxy.ipfsRoundTripper.RoundTrip(req)
+	if err != nil {
+		logger.Error("error making request for header extraction to ipfs: ", err)
+		return err
+	}
+
+	for _, h := range hdrs {
+		dest[h] = res.Header[h]
+	}
+	return nil
+}
+
+// setHeaders sets some headers for all hijacked endpoints:
+// - First we fix CORs headers by making an OPTIONS request to IPFS with the
+//   same Origin. Our objective is to get headers for non-preflight requests
+//   only (the ones we hijack).
+// - Second we add any of the one-time-extracted headers that we deem necessary
+//   or the user needs from IPFS (in case of custom headers), or i by ours.
+//   This may trigger a single POST request to ExtractHeaderPath if they
+//   were not extracted before.
+// - Third we set our own headers.
+func (proxy *Server) setHeaders(dest http.Header, srcRequest *http.Request) {
+	proxy.setCORSHeaders(dest, srcRequest)
+	proxy.setAdditionalIpfsHeaders(dest, srcRequest)
+	proxy.setClusterProxyHeaders(dest, srcRequest)
+}
+
+// see setHeaders
+func (proxy *Server) setCORSHeaders(dest http.Header, srcRequest *http.Request) {
+	// Fix CORS headers by making an OPTIONS request
+
+	// The request URL only has a valid Path(). See http.Request docs.
+	srcURL := fmt.Sprintf("%s%s", proxy.nodeAddr, srcRequest.URL.Path)
+	req, err := http.NewRequest(http.MethodOptions, srcURL, nil)
+	if err != nil { // this should really not happen.
+		logger.Error(err)
+		return
+	}
+
+	req.Header["Origin"] = srcRequest.Header["Origin"]
+	req.Header.Set("Access-Control-Request-Method", srcRequest.Method)
```

Comment:
It appears that we have to set `Access-Control-Request-Headers` also
I am referring https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request for this

Reply:
Good question/catch, but things are ok as they are now.

`Request-Headers` is useful in OPTIONS to get `Allow-Headers` in the response. However `Allow-Headers` is not a header that normal (non OPTIONS) requests should return. The headers we want to extract from OPTIONS are those we can re-use in the subsequent response to a normal request.

Note that a CORS-enabled client may have triggered its own OPTIONS request (which is forwarded to ipfs), so the preflight will work and they will get `Allow-Headers` correctly.  Then they trigger the normal request (which we hijack). We just need to make sure that `Allow-Origin`, `Allow-Credentials` and `Expose-Header` are set, because those are the ones that make sense for normal requests.